README.md
<p align="center">
  <a href="https://github.com/Fliggy-Mobile">
    <img width="200" src="https://gw.alicdn.com/tfs/TB1a288sxD1gK0jSZFKXXcJrVXa-360-360.png">
  </a>
</p>

<h1 align="center">floading</h1>


<div align="center">

<p>Help developers create loading in the most convenient way.</p>

<p>Easily create and manage Loading, support global customization and one-time customization, support timing disappear.</p>

<p><strong>Author：<a href="">Changyi</a>(<a href="fupeng.fp@alibaba-inc.com">fupeng.fp@alibaba-inc.com</a>)</strong></p>

<p>

<a href="https://pub.dev/packages/floading#-readme-tab-">
    <img height="20" src="https://img.shields.io/badge/Version-2.0.0-important.svg">
</a>


<a href="https://github.com/Fliggy-Mobile/floading">
    <img height="20" src="https://img.shields.io/badge/Build-passing-brightgreen.svg">
</a>


<a href="https://github.com/Fliggy-Mobile">
    <img height="20" src="https://img.shields.io/badge/Team-FAT-ffc900.svg">
</a>

<a href="https://www.dartcn.com/">
    <img height="20" src="https://img.shields.io/badge/Language-Dart-blue.svg">
</a>

<a href="https://pub.dev/documentation/floading/latest/floading/floading-library.html">
    <img height="20" src="https://img.shields.io/badge/API-done-yellowgreen.svg">
</a>

<a href="http://www.apache.org/licenses/LICENSE-2.0.txt">
   <img height="20" src="https://img.shields.io/badge/License-Apache--2.0-blueviolet.svg">
</a>

<p>
<p>

</div>

![](https://gw.alicdn.com/tfs/TB1XHHyf8Bh1e4jSZFhXXcC9VXa-1462-552.png)

**English | [简体中文](https://github.com/Fliggy-Mobile/floading/blob/master/README_CN.md)**

> Like it? Please cast your **Star**  🥰 ！

# ✨ Features

- Provide the most convenient way to control **Showing/Hiding** of **Loading**

- Support to modify background color

- Support timing auto hide

- Support local and global custom styles

# 🛠 Guide

## Parameter

```dart
  /// Display Loading
  /// [loading]-custom loading view
  /// [duration]-automatically hide after specified milliseconds. If null, do not hide automatically
  /// [color]-background color when loading, default is [Colors.black54]
  /// [closable]-Is it possible to close loading via the back button
  static show(BuildContext context,
      {Widget loading, int duration, Color color, bool closable = false})


  /// 隐藏 loading。
  /// [context] 有时，开发者可能需要自行传入当前 [context]。
  ///
  /// Hide loading
  /// [context] Sometimes, developers may need to pass in the current [context] by themselves.
  static hide({BuildContext context})
```

## 🌈 Show/hide

![](https://gw.alicdn.com/tfs/TB1PN_dbwgP7K4jSZFqXXamhVXa-720-449.gif)

```dart

/// 显示
///
/// show 
FLoading.show(context);


/// 隐藏
///
/// hide 
FLoading.hide();
```

It's too simple to implement **Show/Hide** of **Loading** through **FLoading**.

## 💎 Custom style

![](https://gw.alicdn.com/tfs/TB1cZHVpcKfxu4jSZPfXXb3dXXa-750-468.gif)

**FLoading** allows developers to freely define the global **Loading** style or the **Loading**
style for a single impression.

### Custom global style

```dart

/// 定义全局 Loading 样式
///
///Define global loading style
FLoading.init(CupertinoActivityIndicator(), backgroundColor: Colors.black38);

/// 显示
///
/// show
FLoading.show(context);
```

### Custom one-shot style

```dart
FLoading.show(context,
  
  /// 通过 loading 配置单次样式
  ///
  /// One-shot style configuration via loading
  loading: Image.asset(
    "assets/loading_gif_2.gif",
    width: 100,
    height: 100,
  ));
```

## 🛰 Timed hiding and background color

![](https://gw.alicdn.com/tfs/TB1LAuKaOpE_u4jSZKbXXbCUVXa-750-468.gif)

```dart
FLoading.show(
  context,
  loading: Image.asset(
    "assets/loading_gif_1.gif",
    width: 200,
    height: 200,
  ),

  /// 定时隐藏
  ///
  /// Timed hiding
  duration: 3000,

  /// 配置背景样式
  ///
  /// Configure background style
  color: Colors.red[300].withOpacity(0.3),
)
```

# 😃 How to use？

Add dependencies in the project `pubspec.yaml` file:

## 🌐 pub dependency

```
dependencies:
  floading: ^<version number>
```

> ⚠️ Attention，please go to [**pub**] (https://pub.dev/packages/floading) to get the latest version number of **FLoading**

## 🖥 Git dependency

```
dependencies:
  floading:
    git:
      url: 'git@github.com:Fliggy-Mobile/floading.git'
      ref: '<Branch number or tag number>'
```

> ⚠️ Attention，please refer to [**FLoading**] (https://github.com/Fliggy-Mobile/floading) official project for branch number or tag.

# 💡 License

```
Copyright 2020-present Fliggy Android Team <alitrip_android@list.alibaba-inc.com>.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at following link.

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

```

### Like it? Please cast your [**Star**](https://github.com/Fliggy-Mobile/floading) 🥰 ！

---

# How to run Demo project?

1. **clone** project to local

2. Enter the project `example` directory and run the following command

```
flutter create .
```

3. Run the demo in `example`





CHANGELOG.md
## 2.0.0
- Support null-safety

## 1.0.2 

- Fix the hiding problem of `FLoading.hide()`

## 1.0.1

- Adaptation **1.17**

## 1.0.0

- the first version


lib/floading.dart
import 'dart:async';

import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';

/// 便捷的创建 loading，支持设置全局样式，也支持设置单次样式。能够自动消失。
///
/// It is convenient to create loading, support setting global style, and also support setting single style. Can disappear automatically.
class FLoading {
  static bool _isShow = false;
  static BuildContext? _cacheContext;
  static Widget? _loading;
  static Color? _backgroundColor;

  static Timer? _timer;

  /// 是否正在显示 loading
  ///
  /// Whether loading is being displayed
  static bool get isShow {
    return _isShow;
  }

  /// 初始化默认的 loading 视图
  ///
  /// Initialize the default loading view
  static init(Widget widget, {Color? backgroundColor}) {
    _loading = widget;
    if (backgroundColor != null) {
      _backgroundColor = backgroundColor;
    }
  }

  /// 显示 Loading
  /// [loading] - 自定义的 Loading 视图
  /// [duration] - 指定毫秒后，自动隐藏。如果为 null，则不自动隐藏
  /// [color] - loading 时的背景颜色，默认为 [Colors.black54]
  /// [closable] - 是否可以通过返回按钮关闭 loading
  ///
  /// Display Loading
  /// [loading]-custom loading view
  /// [duration]-automatically hide after specified milliseconds. If null, do not hide automatically
  /// [color]-background color when loading, default is [Colors.black54]
  /// [closable]-Is it possible to close loading via the back button
  static show(BuildContext context,
      {Widget? loading, int? duration, Color? color, bool closable = false}) {
    if (!_isShow) {
      _isShow = true;
      showGeneralDialog(
        context: context,
        pageBuilder: (BuildContext buildContext, Animation<double> animation,
            Animation<double> secondaryAnimation) {
          final Widget pageChild = Builder(builder: (context) {
            _cacheContext = context;
            Widget widget = loading ?? _loading ?? CupertinoActivityIndicator();
            return WillPopScope(
              onWillPop: () async {
                if (closable) {
                  hide();
                }
                return Future.value(false);
              },
              child: Column(
                mainAxisAlignment: MainAxisAlignment.center,
                children: [widget],
              ),
            );
          });
          return pageChild;
        },
        barrierDismissible: false,
        barrierLabel:
            MaterialLocalizations.of(context).modalBarrierDismissLabel,
        barrierColor: (color ?? _backgroundColor ?? Colors.black38),
        transitionDuration: const Duration(milliseconds: 150),
        transitionBuilder: (BuildContext context, Animation<double> animation,
            Animation<double> secondaryAnimation, Widget child) {
          return FadeTransition(
            opacity: CurvedAnimation(
              parent: animation,
              curve: Curves.easeOut,
            ),
            child: child,
          );
        },
        useRootNavigator: true,
        routeSettings: null,
      );
      if (duration != null) {
        _timer = Timer(Duration(milliseconds: duration), () {
          hide();
        });
      }
    }
  }

  /// 隐藏 loading。
  /// [context] 有时，开发者可能需要自行传入当前 [context]。
  ///
  /// Hide loading
  /// [context] Sometimes, developers may need to pass in the current [context] by themselves.
  static hide({BuildContext? context}) {
    _timer?.cancel();
    _timer = null;
    if(context != null){
      _cacheContext = context;
    }
    if (_isShow && _cacheContext != null) {
      _isShow = false;
      Navigator.pop(_cacheContext!);
      _cacheContext = null;
    }
  }

  /// 出现任何异常时可调用尝试修复
  ///
  /// When any exception occurs, you can call to try to repair
  static fix() {
    hide();
    _cacheContext = null;
    _isShow = false;
  }
}


example/lib/main.dart
import 'package:fbutton/fbutton.dart';
import 'package:fcommon/fcommon.dart';
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:floading/floading.dart';
import 'package:flutter/rendering.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  @override
  void initState() {
    super.initState();
    FLoading.init(CupertinoActivityIndicator(), backgroundColor: Colors.black38);
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Builder(builder: (context) {
        return Scaffold(
          backgroundColor: mainBackgroundColor,
          appBar: AppBar(
            backgroundColor: mainBackgroundColor,
            title: const Text(
              'FLoading',
              style: TextStyle(color: mainTextTitleColor),
            ),
            centerTitle: true,
          ),
          body: SingleChildScrollView(
            child: Container(
              alignment: Alignment.center,
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.center,
                children: [
                  buildTitle("Base FLoading"),
                  buildMiddleMargin(),
                  buildMiddleMargin(),
                  buildMiddleMargin(),

                  /// Base FLoading
                  FButton(
                    text: "Click to show FLoading",
                    style: TextStyle(color: mainTextTitleColor),
                    color: mainBackgroundColor,
                    padding: EdgeInsets.all(12.0),
                    onPressed: () {
                      FLoading.show(context, duration: 2000);
                    },
                    isSupportNeumorphism: true,
                  ),

                  buildMiddleMargin(),
                  buildMiddleMargin(),
                  buildMiddleMargin(),
                  buildTitle("Custom FLoading"),
                  buildMiddleMargin(),
                  buildMiddleMargin(),
                  buildMiddleMargin(),

                  /// Custom FLoading
                  FButton(
                    text: "Click to show FLoading",
                    style: TextStyle(color: mainTextTitleColor),
                    color: mainBackgroundColor,
                    padding: EdgeInsets.all(12.0),
                    onPressed: () {
                      FLoading.show(context,
                          loading: Image.asset(
                            "assets/loading_gif_2.gif",
                            width: 100,
                            height: 100,
                          ),
                          duration: 5000);
                    },
                    isSupportNeumorphism: true,
                  ),

                  buildMiddleMargin(),
                  buildMiddleMargin(),
                  buildMiddleMargin(),
                  buildTitle("Background Color"),
                  buildMiddleMargin(),
                  buildMiddleMargin(),
                  buildMiddleMargin(),

                  FButton(
                    text: "Click to show FLoading",
                    style: TextStyle(color: mainTextTitleColor),
                    color: mainBackgroundColor,
                    padding: EdgeInsets.all(12.0),
                    onPressed: () {
                      FLoading.show(
                        context,
                        loading: Image.asset(
                          "assets/loading_gif_1.gif",
                          width: 200,
                          height: 200,
                        ),
                        duration: 3000,
                        color: Colors.red[300].withOpacity(0.3),
                      );
                    },
                    isSupportNeumorphism: true,
                  ),

                  buildBiggestMargin(),
                  buildBiggestMargin(),

                ],
              ),
            ),
          ),
        );
      }),
    );
  }
}
